# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp
# Commit: ed71fdc2d78d
# Full Hash: ed71fdc2d78dec085a3d153220bba4aeda89992b
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2025-03-05 04:28:59
# Regressor Bug: 1951379
# File Overlap Count: 1
# Description:
#   Bug 1951379 - Part 3: Add NotificationStorageEntry r=asuth
#   
#   This skips handling `vibrate` for now, as we don't ship it and it also complicates test setup.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D240213
# ==============================================================================

diff -r 61ab7888abc4 -r ed71fdc2d78d dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp
--- a/dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp	Tue Mar 04 18:00:17 2025 +0000
+++ b/dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp	Tue Mar 04 18:00:18 2025 +0000
@@ -496,39 +496,27 @@
     return mPromiseHolder.Ensure(__func__);
   }
 
-  NS_IMETHOD Handle(const nsAString& aID, const nsAString& aTitle,
-                    const nsAString& aDir, const nsAString& aLang,
-                    const nsAString& aBody, const nsAString& aTag,
-                    const nsAString& aIcon, const nsAString& aData,
-                    const nsAString& aServiceWorkerRegistrationScope) final {
+  NS_IMETHOD Done(
+      const nsTArray<RefPtr<nsINotificationStorageEntry>>& aEntries) final {
     AssertIsOnMainThread();
-    MOZ_ASSERT(!aID.IsEmpty());
-
-    NotificationDirection dir =
-        StringToEnum<NotificationDirection>(aDir).valueOr(
-            NotificationDirection::Auto);
 
-    // XXX(krosylight): we don't store all notification options
-    IPCNotification notification(
-        nsString(aID),
-        IPCNotificationOptions(nsString(aTitle), dir, nsString(aLang),
-                               nsString(aBody), nsString(aTag), nsString(aIcon),
-                               false, false, nsTArray<uint32_t>(),
-                               nsString(aData)));
+    nsTArray<IPCNotification> notifications(aEntries.Length());
+    for (const auto& entry : aEntries) {
+      auto result = NotificationStorageEntry::ToIPC(*entry);
+      if (result.isErr()) {
+        continue;
+      }
+      MOZ_ASSERT(!result.inspect().id().IsEmpty());
+      notifications.AppendElement(result.unwrap());
+    }
 
-    mNotifications.AppendElement(std::move(notification));
-    return NS_OK;
-  }
-
-  NS_IMETHOD Done() final {
-    mPromiseHolder.Resolve(std::move(mNotifications), __func__);
+    mPromiseHolder.Resolve(std::move(notifications), __func__);
     return NS_OK;
   }
 
  protected:
   virtual ~NotificationsCallback() = default;
 
-  nsTArray<IPCNotification> mNotifications;
   MozPromiseHolder<NotificationsPromise> mPromiseHolder;
 };
 